src/main/java/edu/kit/provideq/toolbox/meta/MetaSolver.java

```diff
@@ -54,12 +53,7 @@ public boolean unregisterSolver(T problemSolver) {
    * @param problem the problem the meta solver is to check its solvers by
    * @return the best suited solver, null in case no suitable solver was found
    */
-  public T findSolver(Problem problem, List<MetaSolverSetting> metaSolverSettings) {
-    Optional<T> solver = solvers.stream()
-        .filter(s -> s.canSolve(problem))
-        .max(Comparator.comparing(s -> s.getSuitability(problem)));
-    return solver.orElse(null);
-  }
+  public abstract T findSolver(Problem problem, List<MetaSolverSetting> metaSolverSettings);
```

Comment:
I think it would be best to have a default implementation that randomizes the choice of the solver. (among those that `canSolve(problem)`). All of our meta solver implementations work the same way atm and it might take some time until we can integrate proper meta solver strategies.

Reply:
Randomizing the solver that is used is what our specific meta solver implementations already do. As a proper implementation is planned in the future, I wouldn't include this default implementation as a design choice as the abstract method urges you to create a strategy which is what the meta solvers are all about.

Reply:
I can't imagine how we can enforce the application of `canSolve` with the current setup, even a template method.
A Meta Solver doesn't get a list of solvers that it can pick from when making its decision (here this would be a list of solvers that return true for `canSolve`). Also, this would add an additional computational overhead if you check `canSolve` for every solver there is before choosing which of the solvers to choose.
You're right that we have a bit of a problem here because we currently need to expect every MetaSolver implementation to validate that their chosen solver can do the job they should do. Maybe we can use a template method to check if the solver that the MetaSolver wants to use can actually be solved via that solver *after* the MetaSolver has made its decision? If it's not valid, we throw an exception and say that there is an implementation error and the solver can't do it. That part isn't very nice, but it's the only thing I can think of right now. If you have any other ideas, please let me know :)

Reply:
I thought about it again and since we've decided to rely on custom algorithms for every meta solver the template method approach doesn't make any sense... I'm not sure we need an exception there, if the solver can't handle a problem it should throw an exception by itself